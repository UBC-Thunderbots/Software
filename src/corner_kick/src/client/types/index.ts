export { IThemeProvider } from './theme';
export { Setting } from './settings';
export { ISettingsCategory, ISettingsItem } from './settingsUI';
export { IRosoutMessage } from './standardROSMessages';
export {
    ILoggerState,
    IRootState,
    IROSState,
    ISettingsState,
    IVisualizerState,
} from './state';
export { IBall, IField } from './visualizer';
