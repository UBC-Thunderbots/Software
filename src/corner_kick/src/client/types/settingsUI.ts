import { Setting } from './settings';

export interface ISettingsCategory {
    path: string;
    title: string;
    icon: string;
    settings: ISettingsItem[];
}

export interface ISettingsItem {
    id: Setting;
    title?: string;
    description?: string;
    type: string;
}
