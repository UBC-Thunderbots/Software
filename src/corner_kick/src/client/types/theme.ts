export interface IThemeProvider {
    colors: IColorProvider;
}

interface IColorProvider {
    fg: string;
    bg: string;

    accent: string;
    subdued: string;
    selected: string;

    panel: string;
    border: string;

    success: string;
    error: string;
    warn: string;
    info: string;
    debug: string;

    red: string;
    orange: string;
    yellow: string;
    green: string;
    cyan: string;
    blue: string;
    purple: string;
}
