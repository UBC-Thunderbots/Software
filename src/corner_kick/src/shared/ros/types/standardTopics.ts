/**
 * @fileoverview Definition file for standard ROS topics.
 * This includes message type that are standard to ROS.
 */

/**
 * @description ROS topic used for logging purposes
 */
export interface IROSOut {
    file: string;
    function: string;
    level: number;
    line: number;
    msg: string;
    name: string;
    topics: string[];
}
