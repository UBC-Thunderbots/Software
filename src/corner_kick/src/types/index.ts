/*
 * This file merges all types in the types folder for easy import
 */

export { Color } from './primitives';
export { IThemeProvider } from './theme';
export { IRosoutMessage } from './standardROSMessages';
export { IRootState, IROSState } from './state';
