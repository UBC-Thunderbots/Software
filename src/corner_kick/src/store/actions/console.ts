/*
 * This file specifies ROS specific action
 *
 * We are using the format specified here
 * @see https://github.com/piotrwitek/typesafe-actions#createaction
 */

import { createAction } from 'typesafe-actions';

import { IRosoutMessage } from 'SRC/types';

/**
 * Sent when a new message is received from a particular topic
 */
export const newRosoutMessage = createAction('console_NEW_ROSOUT', (resolve) => {
    return (message: IRosoutMessage) => resolve({ message });
});
