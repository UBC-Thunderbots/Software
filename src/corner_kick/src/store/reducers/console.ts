/*
 * This file specifies the Console reducer
 */

import { ActionType, getType } from 'typesafe-actions';

import { IMessagesState } from 'SRC/types';

import * as console from '../actions/console';

export type ConsoleAction = ActionType<typeof console>;

const defaultState: IMessagesState = {
    rosout: [],
};

/**
 * Reducer function for Console
 */
export default (state: IMessagesState = defaultState, action: ConsoleAction) => {
    switch (action.type) {
        // Push messages to state if subscribed to /rosout
        case getType(console.newRosoutMessage):
            return {
                ...state,
                rosout: [...state.rosout, action.payload.message],
            };
        default:
            return state;
    }
};
