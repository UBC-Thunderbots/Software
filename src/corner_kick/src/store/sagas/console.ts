/*
 * This file specifies the saga for the Console
 */
import { channel } from 'redux-saga';
import { put, spawn, take, takeLatest } from 'redux-saga/effects';
import { getType } from 'typesafe-actions';

import { TOPIC_ROSOUT, TOPIC_ROSOUT_TYPE } from 'SRC/constants';
import { IRosoutMessage } from 'SRC/types';
import * as ROS from 'SRC/utils/ros';

import { actions } from '../actions';

const consoleChannel = channel();

export default function* init() {
    // Listen to start actions and start Console
    yield takeLatest(getType(actions.ros.connected), startConsole);

    // Start listening to Console messages
    yield spawn(listenToConsoleChannel);
}

/**
 * Take any messages received from Console and push them as Redux actions
 */
function* listenToConsoleChannel() {
    while (true) {
        const action = yield take(consoleChannel);
        yield put(action);
    }
}

/**
 * We subscribe to topic rosout to start receiving messages
 */
function startConsole() {
    ROS.subscribeToROSTopic(
        TOPIC_ROSOUT,
        TOPIC_ROSOUT_TYPE,
        (message: IRosoutMessage) => {
            consoleChannel.put(actions.console.newRosoutMessage(message));
        },
    );
}
