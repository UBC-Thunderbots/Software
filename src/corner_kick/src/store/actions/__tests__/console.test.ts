/**
 * This file is for testing Console actions
 * Each test case has a description of what it tests
 */

import { IRosoutMessage } from 'SRC/types';

import * as consoleActions from '../console';

describe('consoleActions', () => {
    describe('when we receive a new /rosout message', () => {
        test('Dispatches the new /rosout message', () => {
            const mockMessage: IRosoutMessage = {
                file: 'test.ts',
                function: 'test()',
                level: 1,
                line: 123,
                msg: 'Test message',
                name: 'Test',
                topics: ['testTopic'],
            };
            expect(consoleActions.newRosoutMessage(mockMessage)).toMatchSnapshot();
        });
    });
});
