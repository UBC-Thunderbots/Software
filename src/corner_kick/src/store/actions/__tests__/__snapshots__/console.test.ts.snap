// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consoleActions when we receive a new /rosout message Dispatches the new /rosout message 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "message": Object {
      "file": "test.ts",
      "function": "test()",
      "level": 1,
      "line": 123,
      "msg": "Test message",
      "name": "Test",
      "topics": Array [
        "testTopic",
      ],
    },
  },
  "type": "console_NEW_ROSOUT",
}
`;
